src/realm/parser/driver.cpp

```diff
@@ -923,17 +923,41 @@ std::unique_ptr<Subexpr> ConstantNode::visit(ParserDriver* drv, DataType hint)
                         ret = new ConstantStringValue(drv->m_args.string_for_argument(arg_no));
                         break;
                     case type_Binary:
-                        ret = new Value<BinaryData>(drv->m_args.binary_for_argument(arg_no));
+                        ret = new ConstantBinaryValue(drv->m_args.binary_for_argument(arg_no));
                         break;
                     case type_Bool:
                         ret = new Value<Bool>(drv->m_args.bool_for_argument(arg_no));
                         break;
                     case type_Float:
                         ret = new Value<float>(drv->m_args.float_for_argument(arg_no));
                         break;
-                    case type_Double:
-                        ret = new Value<double>(drv->m_args.double_for_argument(arg_no));
+                    case type_Double: {
+                        // In realm-js all number type arguments are returned as double. If we don't cast to the
+                        // expected type, we would in many cases miss the option to use the optimized query node
+                        // instead of the general Compare class.
+                        double val = drv->m_args.double_for_argument(arg_no);
+                        switch (hint) {
+                            case type_Int:
+                            case type_Bool: {
+                                int64_t int_val = int64_t(val);
+                                // Only return an integer if it precisely represents val
+                                if (double(int_val) == val)
+                                    ret = new Value<int64_t>(int_val);
+                                else
+                                    ret = new Value<double>(val);
+                                break;
+                            }
+                            case type_Float:
+                                ret = new Value<float>(float(val));
+                                break;
+                            case type_Double:
+                                ret = new Value<double>(val);
+                                break;
+                            default:
+                                break;
```

Comment:
I suggest the default case should always be a double value, this will handle mixed, double, decimal128 and any future numeric comparisons.
```suggestion
                            default:
                                ret = new Value<double>(val);
                                break;
```